### ISSUE TITLE
Feature: emscripten_get_callstack with DWARF info

### ISSUE THREAD

**Issue by yowl** [emscripten-core/emscripten#12113]
The flags parameter I don't believe has an option to get filenames from the DWARF info, I've tried 0 and 8.  Could a flag be added that retrieves the file information from DWARF as it does for source maps?

**Comment by kripken** [comment]
It sounds like that would require shipping a DWARF parser - seems pretty heavyweight?

**Comment by yowl** [comment]
I suppose it would, I hadn't thought of the practical difference between this and the way it works with source maps.  If its too heavy feel free to close.

**Comment by yowl** [comment]
Maybe with Wasm exceptions, there will be some sort of libunwind that will help?  https://github.com/WebAssembly/tool-conventions/blob/master/EHScheme.md

**Comment by kripken** [comment]
cc @aheejin

**Comment by aheejin** [comment]
Platforms that use Dwarf CFI can use libunwind to do that, but wasm unwinding is done by VM and our libunwind is just a stub that triggers a VM to start unwinding process, so I'm not very sure. But if we later support two-phase unwinding, and if we have the Dwarf info in our binary, I wonder the VM or devtools can possibly use that info to show the stack traces? @RReverser @backes

**Comment by RReverser** [comment]
I think that might be doable, although in our current model DWARF integration sits on a different level and might be hard to integrate with `Error.prototype.stack`? cc @bmeurer

**Comment by bmeurer** [comment]
Couldn't this be implemented similarly to `backtrace(3)` on Linux? I haven't looked how that is implemented tho, but I suspect this has to be able to deal with debug info in some (lightweight) way.

Integrating this with the DevTools doesn't sound like a viable approach to me, as we anyways need a fallback solution, and `emscripten_get_callstack` is probably most useful exactly when you don't have the DevTools open (aka in production).

**Comment by RReverser** [comment]
> Integrating this with the DevTools doesn't sound like a viable approach to me, as we anyways need a fallback solution

Yeah I don't think it has to be DevTools-specific, but rather something integrated with `Error.prototype.stack` as mentioned above. Could be good to think about whether we can propagate DWARF to those stacks in the future, as it could be useful for many error reporting services.

**Comment by bmeurer** [comment]
DevTools does post-processing on stack traces, which are printed to the console. So even with source maps you'll get an exhanced version of the stack trace printed when DevTools is open. But TBH I don't really like this behavior, I'd be in favor of having this as part of the engine / program instead and completely independent of DevTools.

**Comment by backes** [comment]
> I think that might be doable, although in our current model DWARF integration sits on a different level and might be hard to integrate with `Error.prototype.stack`? cc @bmeurer

+1, V8 current does not parse any DWARF at all, that's all done in DevTools. So the better model is probably that the VM still provides a low-level stack trace, and either DevTools or the toolchain translates that to a source-level stack trace, using DWARF or any other available debugging information.

**Comment by aheejin** [comment]
I think this feature will be still useful, even though we don't have a near time plan for this. Commenting to make this stay open.
